### ISSUE TITLE
Log loss metric during training does not match scikit-learn log_loss of the trained model

### ISSUE THREAD

**Issue by stefanhgm** [interpretml/interpret#247]
Hi,

I wanted to validate the log loss results (for classification) that are displayed during training and I found that they are different to log_loss in scikit-learn. I created a minimal working example with a known validation set for easier debugging.

```
x_train = np.random.rand(50, 10)
y_train = np.random.randint(2, size=50)
x_valid = np.random.rand(10, 10)
y_valid = np.array([0, 1, 0, 1, 0, 1, 0, 1, 0, 1])

model = ExplainableBoostingClassifier(outer_bags=1, validation_size=10, n_jobs=1, interactions=0)
x = np.concatenate((x_train, x_valid), axis=0)
y = np.concatenate((y_train, y_valid), axis=0)
model.fit(x, y)
print(log_loss(y_valid, model.predict_proba(x_valid)))
```
I also changed `ebm_train_test_split` in `utils.py` to use the last n=10 element as X_val / y_val. I also verified this by hand.

```
[...]
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=test_size, random_state=(random_state - (-2147483648)) if random_state < 0 else random_state, shuffle=False, stratify=None)
[...]
```

However, during training the logging shows:

`interpret.glassbox.ebm.internal - INFO - End boosting Main, Best Metric: 0.5814300537109375, Num Rounds: 137`

while scikit-learn's log_loss gives `0.6246216731428145`.

I already looked into EbmStats.h to identify any differences in the calculation, but it's very hard to do that without a proper debugger. Shouldn't the results be the same or is there anything I am missing?

Thanks!

**Comment by interpret-ml** [comment]
Hi @stefanhgm --

As you've surmised, that metric isn't exactly a log loss value, so we're careful not to identify what it is in the package besides calling it a "metric".  To improve the speed of the package, we use approximations for calculating the log loss based on this paper ( http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.9.4508&rep=rep1&type=pdf )

Internally we use log loss only for early stopping, and for early stopping we only care that the relative difference between boosting rounds is roughly correct, so this approximation works out nicely for us.  We've benchmarked the use of this approximation internally and found no detectable degradation compared to using the exact log loss calculation.  If you want to compute the true log loss we recommend using sklearn to calculate it afterwards as you've done above.  Someday in the future we'll add support for calculating the true log loss, and many other metrics, but given you can already do this with sklearn it's less of a priority than many other features in our backlog.

If you'd like to disable the use of approximate floating point calculations, this can be done by commenting out these two lines and then recompiling the C++.  If you make this change then our reported internal metric value should be equivalent to the sklearn computed log loss:

https://github.com/interpretml/interpret/blob/15c72f80cf9f7c6e6aa3f385a5404ec8b9cf60cb/shared/ebm_native/common_c/common_c.h#L168

-InterpretML team
